File: java/mini-ct/src/main/java/ControlTower.java
import akka.NotUsed;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.http.javadsl.ConnectHttp;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.stream.ActorMaterializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Sink;
import api.ControlTowerApi;
import domain.GPSUpdate;
import kafka.KafkaGPSConsumer;
import updates.GPSUpdatesActor;

public class ControlTower {
    private static final int PORT = 9081;
    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
    private static final String TOPIC = "jfall-gps-updates";

    public static void main(String args[]) {

        System.out.println("Starting the Control Tower");

        // Default Akka setup
        ActorSystem system = ActorSystem.create("routes");
        final ActorMaterializer materializer = ActorMaterializer.create(system);
        final Http http = Http.get(system);

        // Create an actor that keeps track of the latest updates per device
        Props actorProps = Props.create(GPSUpdatesActor.class);
        ActorRef gpsUpdatesActor = system.actorOf(actorProps);

        // Create a Kafka consumer that reads messages from Kafka and sends them to the actor
        KafkaGPSConsumer consumer = new KafkaGPSConsumer(system, BOOTSTRAP_SERVERS, TOPIC);
        Sink<GPSUpdate, NotUsed> sink = Sink.actorRef(gpsUpdatesActor, "done");
        consumer.createSource().to(sink).run(materializer);

        // Create an endpoint so you can retrieve the latest updates per device
        System.out.println("Start endpoint on port " + PORT);
        ControlTowerApi api = new ControlTowerApi(gpsUpdatesActor);
        final Flow<HttpRequest, HttpResponse, NotUsed> routeFlow = api.getRoute().flow(system, materializer);
        http.bindAndHandle(routeFlow, ConnectHttp.toHost("localhost", PORT), materializer);
    }
}

File: mini-ct/src/main/scala/mini_ct/ControlTower.scala
package mini_ct

import akka.actor.{ActorRef, ActorSystem}
import akka.http.scaladsl.Http
import akka.stream.scaladsl.Sink
import akka.stream.{ActorMaterializer, Materializer}
import mini_ct.api.ControlTowerApi
import mini_ct.kafka.GPSUpdateConsumer
import mini_ct.updates.GPSUpdatesActor

import scala.concurrent.ExecutionContext
import scala.util.Properties

object ControlTower extends App {

  import GPSUpdatesActor._

  val port = 8888
  val bootstrapServers = Properties.envOrElse("BOOTSTRAP_URL", "localhost:9092")
  val topic = if(args.length == 1) args(0) else Properties.envOrElse("TOPIC", "jfall-gps-updates")

  println("Starting the Control Tower")

  // Some setup to get Akka HTTP up and running
  implicit val system: ActorSystem = ActorSystem()
  implicit val executionContext: ExecutionContext = system.dispatcher
  implicit val mat: Materializer = ActorMaterializer()

  // Actor
  val supervisorActor: ActorRef = system.actorOf(GPSUpdatesActor.props)

  // Kafka
  val GPSUpdateConsumer = new GPSUpdateConsumer(system, bootstrapServers, topic)
  val supervisorSink = Sink.actorRefWithAck(supervisorActor, Init, Ack, Complete)

  startApplication()
  createEndpoints()

  /**
    * Start reading from Kafka and process every sensor update
    */
  def startApplication() = {
    GPSUpdateConsumer.source
      .to(supervisorSink)
      .run
  }

  /**
    * Create endpoints for this application so you can retrieve updates
    */
  def createEndpoints()= {
    println("Start endpoint")
    val api = new ControlTowerApi(supervisorActor)
    Http().bindAndHandle(
      handler = api.route,
      interface = "0.0.0.0",
      port = port
    )
  }
}
